src/rpc/methods/f3/types.rs

```diff
@@ -135,3 +135,102 @@ impl PartialOrd for F3PowerEntry {
         Some(self.cmp(other))
     }
 }
+
+#[derive(Debug, Default)]
+pub struct F3LeaseManager(RwLock<HashMap<u64, chrono::DateTime<chrono::Utc>>>);
+
+impl F3LeaseManager {
+    pub fn get_active_participants(&self) -> HashSet<u64> {
+        let now = chrono::Utc::now();
+        self.0
+            .read()
+            .iter()
+            .filter_map(|(id, expire)| if expire > &now { Some(*id) } else { None })
+            .collect()
+    }
+
+    pub fn upsert_defensive(
+        &self,
+        id: u64,
+        new_lease_expiration: chrono::DateTime<chrono::Utc>,
+        old_lease_expiration: chrono::DateTime<chrono::Utc>,
+    ) -> anyhow::Result<bool> {
+        if new_lease_expiration > chrono::Utc::now() + chrono::Duration::minutes(5) {
+            anyhow::bail!("F3 participation lease cannot be over 5 mins");
+        } else if new_lease_expiration < chrono::Utc::now() {
+            anyhow::bail!("F3 participation lease is in the past");
+        }
+
+        // if the old lease is expired just insert a new one
+        if old_lease_expiration < chrono::Utc::now() {
```

Comment:
let's declare and use a single `now`. If we don't, we risk some weird conditions happening. Let's also document it, it's not a trivial name.

Reply:
Fixed.